### PR TITLE
MM-41323: Secure plugin marketplace modification

### PR DESCRIPTION
**Pull request by agnivade**
We disable plugin marketplace modification if plugin uploads
are disabled

https://mattermost.atlassian.net/browse/MM-41323
```release-note
NONE
```

### CODE REVIEW

File: api4/config.go [hunk 1]
```diff
@@ -145,6 +145,11 @@ func updateConfig(c *Context, w http.ResponseWriter, r *http.Request) {
 	// Do not allow certificates to be changed through the API
 	cfg.PluginSettings.SignaturePublicKeyFiles = appCfg.PluginSettings.SignaturePublicKeyFiles
 
+	// Do not allow marketplace URL to be toggled through the API if EnableUploads are disabled.
+	if cfg.PluginSettings.EnableUploads != nil && !*cfg.PluginSettings.EnableUploads {
+		cfg.PluginSettings.MarketplaceURL = appCfg.PluginSettings.MarketplaceURL
```
Comment (isacikgoz) [inline api4/config.go:150]: We should assign values right?
```suggestion
		*cfg.PluginSettings.MarketplaceURL = *appCfg.PluginSettings.MarketplaceURL
```

File: api4/config.go [hunk 2]
```diff
@@ -260,6 +265,15 @@ func patchConfig(c *Context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// Do not allow marketplace URL to be toggled if plugin uploads are disabled.
+	if cfg.PluginSettings.MarketplaceURL != nil && cfg.PluginSettings.EnableUploads != nil {
+		// Breaking it down to 2 conditions to make it simple.
+		if *cfg.PluginSettings.MarketplaceURL != *appCfg.PluginSettings.MarketplaceURL && !*cfg.PluginSettings.EnableUploads {
```
Comment (isacikgoz) [inline api4/config.go:271]: Shouldn't we check the `appCfg.PluginSettings.EnableUploads` as it is the source of truth?
```suggestion
		if *cfg.PluginSettings.MarketplaceURL != *appCfg.PluginSettings.MarketplaceURL && !*appCfg.PluginSettings.EnableUploads {
```

File: api4/config.go (hunk 1 above)
Comment (agnivade) [reply api4/config.go:150]: Yes, the earlier lines also copied the pointer, which is risky. I'll change all of them.

File: api4/config.go (hunk 2 above)
Comment (agnivade) [reply api4/config.go:271]: The value already gets replaced in L143, but you are right. Using `appCfg` makes it clear.

**Review by srkgupta — CHANGES_REQUESTED** [review]
Hi @agnivade 

Observed some issues:

1) Currently the config `Enable Remote Marketplace:` is not disabled when `Upload Plugin` option is disabled. Can you please also disable this option since a user cannot do anything as the Marketplace URL field is already disabled with this PR. 

2) Can we display a similar message when the Marketplace URL field is disabled:
`Enable plugin uploads in config.json.`

3) This also requires the following doc to be updated:
https://developers.mattermost.com/integrate/admin-guide/admin-plugins-beta/#custom-plugins

**Comment by agnivade** [comment]
Makes sense. Agree with all of those.

**Comment by agnivade** [comment]
PTAL along with the new webapp branch.

**Review by srkgupta — APPROVED** [review]
Thanks @agnivade. Changes LGTM

**Review by isacikgoz — APPROVED** [review]

**Pull request merged**